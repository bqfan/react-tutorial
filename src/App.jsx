import { useEffect, useState } from "react";
import { UserDetails } from "./components/UserDetails";
// import { PostContainer } from "./components/PostContainer";
// import { UserContext } from "./utils/contexts/UserContext";
// import { useFetchUser } from "./utils/hooks/useFetchUser";
// import { Outlet, Link, useNavigate, Navigate } from 'react-router-dom';


function App({ usersData }) {
    // const { user, loading, error } = useFetchUser(1);
    // console.log(user, loading, error);

    // const [userData, setUserData] = useState({

    // });

    // const navigate = useNavigate();

    // const [users, setUsers] = useState([
    //     {
    //         id: 1,
    //         username: "user1",
    //         email: "user1@example.com"
    //     },
    //     {
    //         id: 2,
    //         username: "user2",
    //         email: "user2@example.com"
    //     }
    // ]);
    const [users, setUsers] = useState(usersData);
    // useEffect(() => {
    //     if (!loading && !error && user) {
    //         setUserData(user);
    //         // navigate('/users');
    //     }
    // // }, [loading, error, user, navigate]);
    //     }, [loading, error, user]);
    return (
        <>
            {users.map((user) => (<UserDetails key={user.id} user={user} setUsers={setUsers} />))}
            {/* <nav>
                <ul>
                    <li>
                        <Link to="/">Home</Link>
                    </li>
                    <li>
                        <Link to="/users">Users</Link>
                    </li>
                    <li>
                        <Link to="/blogposts">Blogs</Link>
                    </li>
                </ul>
            </nav>
            <div>
                <label htmlFor="data">Enter data</label>
                <input type="text" id="data" onChange={(e) => {
                    console.log(e.target.value);
                    if (e.target.value.length > 10) {
                        navigate("/blogposts", {
                            state: {
                                posts: [
                                    {
                                        id: 1,
                                        title: "Hello World",
                                        content: "Welcome to my first post"
                                    },
                                ]
                            }
                        })
                    }
                }} />
            </div>
            <UserContext.Provider value={{ ...userData, setUserData }}>
                <div>
                    { loading ? "Loaing ..." : <PostContainer /> }
                </div>
            </UserContext.Provider>
            <Outlet /> */}
        </>
    )
}

export default App;