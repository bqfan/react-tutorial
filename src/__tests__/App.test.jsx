import { expect, it, describe } from 'vitest';
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { within } from "@testing-library/dom";
import App from "../App";

describe('when there is only 1 user', () => {
    it('should render users', () => {
        const { container } = render(<App usersData={[]} />);
        expect(container).toMatchSnapshot();
    });

    describe('Edit Button is clicked and display save button', () => {
        it('should render save button', async () => {
            render(<App usersData={[
                {
                    id: 1,
                    username: "user1",
                    email: "user1@example.com"
                },
            ]} />);
            // console.log(screen.getByText("Edit"));
            const editButton = screen.getByRole('button', { name: "Edit" });
            await userEvent.click(editButton);
            const saveButton = screen.getByRole('button', { name: "Save" });
            expect(saveButton).toBeInTheDocument();
        });
    
        it('should display username & email input fields', async () => {
            render(<App  usersData={[
                {
                    id: 1,
                    username: "user1",
                    email: "user1@example.com"
                },
            ]} />);
            const editButton = screen.getByRole('button', { name: "Edit" });
            await userEvent.click(editButton);
            // expect(screen.getByRole('textbox', { name: 'username' })).toBeInTheDocument();
            // expect(screen.getByRole('textbox', { name: 'email' })).toBeInTheDocument();
            expect(screen.getByLabelText('Username:')).toBeInTheDocument();
            expect(screen.getByLabelText('Email:')).toBeInTheDocument();
        });
    })
});

describe('when there is only 2 user', () => {
    it('should have two users', () => {
        render(<App usersData={[
            {
                id: 1,
                username: "user1",
                email: "user1@example.com"
            },
            {
                id: 2,
                username: "user2",
                email: "user2@example.com"
            },
        ]} />);

        expect(screen.getByText('user1')).toBeInTheDocument();
        expect(screen.getByText('user2')).toBeInTheDocument();
    })

    it('should click edit button for 1st user', async () => {
        render(<App usersData={[
            {
                id: 1,
                username: "user1",
                email: "user1@example.com"
            },
            {
                id: 2,
                username: "user2",
                email: "user2@example.com"
            },
        ]} />);

        const userDetails = screen.getByTestId('user-details-1');
        expect(within(userDetails).getByText('user1')).toBeInTheDocument();
        expect(within(userDetails).queryByText('user2')).toBeNull();
        const editBtn = within(userDetails).getByRole('button', { name: "Edit" });
        await userEvent.click(editBtn);
        expect(within(userDetails).getByRole("button", { name: 'Save' })).toBeInTheDocument;
    })

    it('should edit 2nd username and save', async () => {
        render(<App usersData={[
            {
                id: 1,
                username: "user1",
                email: "user1@example.com"
            },
            {
                id: 2,
                username: "user2",
                email: "user2@example.com"
            },
        ]} />);
        
        const userDetails = screen.getByTestId('user-details-2');
        expect(within(userDetails).getByText('user2')).toBeInTheDocument();
        expect(within(userDetails).queryByText('user1')).toBeNull();

        await userEvent.click(
            within(userDetails).getByRole('button', { name: "Edit" })
        );

        await userEvent.type(
            within(userDetails).getByLabelText('Username:'),
            'edited'
        );

        await userEvent.click(
            within(userDetails).getByRole('button', { name: 'Save' })
        );

        expect(
            within(userDetails).queryByLabelText('Username:')
        ).toBeNull();

        expect(
            within(userDetails).getByText('user2edited')
        ).toBeInTheDocument();
    })
});

describe('Updating UserContext', () => {
    it('should update displayName', async () => {
        render(<App usersData={[]} />);

        await userEvent.type(
            screen.getByLabelText("Update Name:"),
            "Updated Display Name"
        );
        
        await userEvent.click(
            screen.getByRole("button", "Update Display Name")
        );

        expect(screen.getByText("Display Name: Updated Display Name")).toBeInTheDocument();
    })
})
