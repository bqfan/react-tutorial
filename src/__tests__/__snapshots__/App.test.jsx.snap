// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App > should render users 1`] = `<div />`;

exports[`when there is only 1 user > should render users 1`] = `<div />`;
